aineko/core/dataset.py

```diff
@@ -132,30 +140,99 @@ def _validate_message(
+        # Initialize consumers if not already initialized by polling
+        while not partitions:
+            self.consumer.poll(timeout=0)
+            partitions = self.consumer.assignment()
+
+        for partition in partitions:
+            partition.offset = (
+                self.consumer.get_watermark_offsets(
+                    partition, cached=self.cached
+                )[1]
+                - 1
+            )
+
+        self.consumer.assign(partitions)
+
     def consume(
         self,
-        how: str = "next",
-        timeout: Optional[int] = None,
+        how: Literal["next", "last"] = "next",
+        timeout: Optional[float] = None,
     ) -> Optional[dict]:
-        """Reads a message from the dataset.
+        """Polls a message from the dataset.
 
         Args:
-            how: how to read the message
+            how: how to read the message.
                 "next": read the next message in the queue
+                "last": read the last message in the queue
+            timeout: seconds to poll for a resopnse from kafka broker.
+            If using how="last",
 
         Returns:
-            msg: message from the dataset
+            message from the dataset
         """
+        if how not in ["next", "last"]:
+            raise ValueError(f"Invalid how: {how}. Expected `next` or `last`.")
+
         timeout = timeout or self.kafka_config.get("CONSUMER_TIMEOUT")
         if how == "next":
             # next unread message from queue
-            message = self._validate_message(
-                self.consumer.poll(timeout=timeout)
-            )
-        else:
-            raise ValueError(f"Invalid how: {how}. Expected 'next'.")
+            message = self.consumer.poll(timeout=timeout)
+
+        if how == "last":
+            # last message from queue
+            self._update_offset_to_latest()
+            message = self.consumer.poll(timeout=timeout)
+
+        self.cached = True
+
+        return self._validate_message(message)
+
+    def _consume_message(
+        self, how: Literal["next", "last"], timeout: Optional[float] = None
+    ) -> dict:
+        """Calls the consume method and blocks until a message is returned.
+
+        Args:
+            how: See `consume` method for available options.
+
+        Returns:
+            message from dataset
+        """
+        while True:
+            try:
+                message = self.consume(how=how, timeout=timeout)
+                if message is not None:
+                    return message
+            except KafkaError as e:
+                if e.code() == "_MAX_POLL_EXCEEDED":
+                    continue
+                raise e
+
+    def next(self) -> dict:
+        """Wraps `consume(how="next")`, blocks until available.
+
+        Returns:
+            msg: message from the dataset
+        """
+        return self._consume_message(how="next")
 
-        return message
+    def last(self, timeout: int = 1) -> dict:
```

Comment:
> Note that a non-zero value for timeout is required if consuming the last message since we update the offset on every iteration. Set the timeout to the desired maximum latency of the node.

Based on this, I think we should raise a `ValueError` or at least document in-line that `timeout` needs to be non-zero